libs/dmlf/src/muddle2_learner_networker.cpp

```diff
@@ -0,0 +1,186 @@
+                                                 MuddleChannel                          channel_tmp)
+  : channel_tmp_{channel_tmp}
+{
+  json::JSONDocument doc{cloud_config};
+
+  FETCH_LOG_INFO("Muddle2LearnerNetworker", "here 1");
+  if (netm)
+  {
+    this->netm = netm;
+  }
+  else
+  {
+    this->netm = std::make_shared<NetworkManager>("NetMgrA", 4);
+  }
+  this->netm->Start();
+
+  auto my_config = doc.root()["peers"][instance_number];
+  auto self_uri  = Uri(my_config["uri"].As<std::string>());
+  auto port      = self_uri.GetTcpPeer().port();
+  auto privkey   = my_config["key"].As<std::string>();
+
+  // ident = CreateIdentity();
+  ident = LoadIdentity(privkey);
+
+  auto addr = self_uri.GetTcpPeer().address();
+
+  mud = muddle::CreateMuddle("Test", ident, *(this->netm), addr);
+  mud->SetPeerSelectionMode(muddle::PeerSelectionMode::KADEMLIA);
+
+  std::unordered_set<std::string> initial_peers;
+  if (instance_number > 0)
+  {
+    initial_peers.insert(doc.root()["peers"][0]["uri"].As<std::string>());
+  }
+
+  mud->Start(initial_peers, {port});
+
+  server = std::make_shared<Server>(mud->GetEndpoint(), 1, 1);
+  proto  = std::make_shared<Muddle2LearnerNetworkerProtocol>(*this);
+
+  server->Add(1, proto.get());
+
+  auto config_peers = doc.root()["peers"];
+
+  for (std::size_t peer_number = 0; peer_number < config_peers.size(); peer_number++)
+  {
+    if (peer_number != instance_number)
+    {
+      peers.push_back(config_peers[peer_number]["pub"].As<std::string>());
+    }
+  }
+}
+
+// TOFIX remove return value
+uint64_t Muddle2LearnerNetworker::RecvBytes(const byte_array::ByteArray &b)
+{
+  switch (channel_tmp_)
+  {
+  case MuddleChannel::DEFAULT:
+    AbstractLearnerNetworker::NewMessage(b);
+    break;
+  case MuddleChannel::MULTIPLEX:
+    AbstractLearnerNetworker::NewDmlfMessage(b);
+    break;
+  default:;
+  }
+  return 0;
+}
+
+Muddle2LearnerNetworker::~Muddle2LearnerNetworker() = default;
+
+void Muddle2LearnerNetworker::pushUpdate(const std::shared_ptr<UpdateInterface> &update)
+{
+  auto client = std::make_shared<RpcClient>("Client", mud->GetEndpoint(), 1, 1);
+  auto data   = update->serialise();
+
+  PromiseList promises;
+  promises.reserve(20);
+
+#pragma clang diagnostic ignored "-Wunused-variable"
+  for (const auto &target_peer : peers)
+  {
+    promises.push_back(client->CallSpecificAddress(
+        fetch::byte_array::FromBase64(byte_array::ConstByteArray(target_peer)), 1, 1, data));
+  }
+
+  for (auto &prom : promises)
+  {
+    prom->Wait();
+  }
+}
+
+void Muddle2LearnerNetworker::pushUpdateType(const std::string &                     type,
+                                             const std::shared_ptr<UpdateInterface> &update)
+{
+  auto client = std::make_shared<RpcClient>("Client", mud->GetEndpoint(), 1, 1);
+  auto data   = update->serialise(type);
+
+  PromiseList promises;
+  promises.reserve(20);
+
+#pragma clang diagnostic ignored "-Wunused-variable"
```

Comment:
Not needed?

Reply:
Not even sure what it's doing there...
done

Reply:
edleted